askama_shared/src/parser.rs

```diff
@@ -879,24 +902,28 @@ fn block_for<'a>(i: &'a [u8], s: &'a Syntax<'a>) -> IResult<&'a [u8], Node<'a>>
                     cut(tuple((
                         |i| tag_block_start(i, s),
                         opt(tag("-")),
+                        opt(|i| else_block(i, s)),
                         ws(tag("endfor")),
                         opt(tag("-")),
                     ))),
                 ))),
             ))),
         ))),
     ));
-    let (i, (pws1, _, (var, _, (iter, nws1, _, (block, (_, pws2, _, nws2)))))) = p(i)?;
-    Ok((
-        i,
-        Node::Loop(
-            Ws(pws1.is_some(), nws1.is_some()),
-            var,
-            iter,
-            block,
-            Ws(pws2.is_some(), nws2.is_some()),
-        ),
-    ))
+    let (i, (pws1, _, (var, _, (iter, nws1, _, (block, (_, pws2, else_block, _, nws2)))))) = p(i)?;
+    let ws1 = Ws(pws1.is_some(), nws1.is_some());
+    match else_block {
```

Comment:
Here, it would be nice to keep the `match` block more tightly scoped, having it return the right part of `ws2`, the left part of `ws3`, and the else block. Makes it a bit clearer what's going and reduces code duplication.

Reply:
Would it make sense to add helper function like this?

```r
fn parse_ws(inner: impl Fn(Input) -> Output) -> impl Fn(Input) -> IResult<Input, (Ws, Output)>
```

Matching the optional hyphens and returning  `Ws(pws.is_some(), nws.is_some())` comes up _a lot_.

Reply:
I just noticed that I can use unwrap_or_default() to that no match at all is needed.